Feature/jpaspec

PaleRoom opened the pull request.

Commit:
Added Search Feature

Commit:
Added PDFbox pdf-creating method

Commit:
Added PDFbox pdf-creating method
Exchange method

Commit:
Added Saving to XLSX Service

PaleRoom committed:
Deleted useless classes

PaleRoom committed:
Merge branch 'develop' into feature/jpaspec

# Conflicts:
#	pom.xml
#	src/main/java/ru/ncs/DemoShop/controller/ProductController.java
#	src/main/java/ru/ncs/DemoShop/controller/ProductControllerImpl.java
#	src/main/java/ru/ncs/DemoShop/controller/converter/ConvertToImmutableCreateProductRequest.java
#	src/main/java/ru/ncs/DemoShop/controller/converter/ConvertToImmutableUpdateProductRequest.java
#	src/main/java/ru/ncs/DemoShop/controller/request/CreateProductRequest.java
#	src/main/java/ru/ncs/DemoShop/controller/request/UpdateProductRequest.java
#	src/main/java/ru/ncs/DemoShop/controller/response/GetProductResponse.java
#	src/main/java/ru/ncs/DemoShop/model/Product.java
#	src/main/java/ru/ncs/DemoShop/repository/ProductRepository.java
#	src/main/java/ru/ncs/DemoShop/service/ProductService.java
#	src/main/java/ru/ncs/DemoShop/service/ProductServiceImpl.java
#	src/main/java/ru/ncs/DemoShop/service/converter/ConvertToProductUpdate.java
#	src/main/java/ru/ncs/DemoShop/service/data/ProductDTO.java
#	src/main/java/ru/ncs/DemoShop/service/immutable/ImmutableCreateProductRequest.java
#	src/main/java/ru/ncs/DemoShop/service/immutable/ImmutableUpdateProductRequest.java

PaleRoom committed:
-Develop Merged in Jpaspec
-pom.xml fixed

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/ProductController.java:
```diff
@@ -35,4 +39,6 @@ UUID updateProduct(@PathVariable("id") UUID id,
     void deleteProduct(@PathVariable("id") UUID id);
 
+    @PostMapping("/search")
+    GetListResponse searchProducts(SearchProductRequest searchProductRequest) throws IOException;
```
Пусть метод возвращает просто List ProductResponse. Класс GetListResponse не нужен и связанный с ним конвертер. Метод getProducts() пусть тоже возвращает List ProductResponse. Конвертеры ConvertToGetListResponse и ConvertToResponseList не нужны + у них стремные имена) Посмотришь на них - и ниче не понятно) И там и там лист..)

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/ProductControllerImpl.java:
```diff
@@ -27,16 +32,23 @@
     public GetListResponse getProducts() {
-        List<GetProductResponse> list = conversionService.convert(productService.findAll(), List.class);
-        return conversionService.convert(list, GetListResponse.class);
+        List<ProductDTO> ListDTO = productService.findAll();
```
- [x] Давай все это стримом. List GetProductResponse list =productService.findAll().stream().map(dto -> conversionService.convert(dto, GetProductResponse.class).collect(Collectors::toList);
Метод будет возвращать List GetProductResponse

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/ProductControllerImpl.java:
```diff
@@ -62,4 +74,22 @@ public void deleteProduct(@PathVariable("id") UUID id) {
+                conversionService.convert(searchProductRequest, ImmutableSearchProductRequest.class);
+        List<ProductDTO> ListDTO = productService.searchProducts(immutableSearchProductRequest);
+
+        List<GetProductResponse> listGPR = new ArrayList<>();
```
Тоже самое. Сделай через стрим. Все обработки коллекций через стрим.

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/SearchedToPDFsaver.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.List;
+
+@Component
+public class SearchedToPDFsaver {
```
Оба сейвера документов вынести в один класс в пакете сервис. В пакете контроллера их точно не должно быть. Ну и вызываться они не должны из контроллера. В сервисе же есть отдельный метод для поиска по спецификации. вот там и давай их вызовем

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/SearchedToPDFsaver.java:
```diff
@@ -0,0 +1,65 @@
+
+@Component
+public class SearchedToPDFsaver {
+    public void saveSearchedToPdf(GetListResponse listResponse) throws IOException {
```
Методы должны работать с List ProductDto, а не с сущностями контроллера

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/SearchedToXlsSaver.java:
```diff
@@ -0,0 +1,88 @@
+    private XSSFSheet sheet;
+    private XSSFWorkbook workbook;
+
+    public void SaveSearchedToXls(GetListResponse listResponse) throws IOException {
```
аналогично, принимать должны List ProductDto

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/request/SearchProductRequest.java:
```diff
@@ -0,0 +1,18 @@
+@Builder
+@Data
+public class SearchProductRequest {
+    String name;
```
поля private

GUR-ok commented on src/main/java/ru/ncs/DemoShop/repository/specification/ProductSpecification.java:
```diff
@@ -0,0 +1,35 @@
+
+    public static Specification<Product> likeName(String name) {
+        if (name == null) {
+            return null;
```
Возвращать нулл - это работает? Может нужно возвращать Specification.where(null)?

PaleRoom replied:
Сделал через criteriaBuilder.conjunction();

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/ProductServiceImpl.java:
```diff
@@ -130,4 +133,21 @@ public void increasePrice(double mod) throws InterruptedException {
+
+        List<Product> foundList = productRepository.findAll(specs);
+
+        List<ProductDTO> listDTO = new ArrayList<>();
```
Обработать стримом

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/immutable/ImmutableSearchProductRequest.java:
```diff
@@ -0,0 +1,16 @@
+@Value
+public class ImmutableSearchProductRequest {
+
+    String name;
```
Тут все ок - private не нужно, т.к. value. Value дает private final. Поэтому и получается иммутабельный класс

GUR-ok commented on src/main/resources/db/changelog/add-availability-column.xml:
```diff
@@ -0,0 +1,14 @@
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+    <!-- Здесь мы позже напишем скрипт -->
+    <property name="now" value="now()" dbms="postgresql"/>
+
```
Добавить preCondition, что такая таблица существует

PaleRoom replied:
Добавил во всех скриптах. При создании таблицы - проверяет ее отсутствие.

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/converter/ConvertToResponseList.java:
```diff
@@ -21,6 +21,5 @@ public List<GetProductResponse> convert(List<ProductDTO> source) {
             list.add(conversionService.convert(product, GetProductResponse.class));
```
Давай вообще этот конвертер уберем. У тебя есть конвертер для одиночного объекта ConvertToGetProductResponse (кстати лучше его назвать по-другому - ProductDtoToGetProductResponseConverter, а то у тебя непонятно по названию что во что конвертится). А преобразования списка в список делать стримом находу, вызывая конвертер единичного объекта. Конвертить список в список конвертером - плохая практика.

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/ProductControllerImpl.java:
```diff
@@ -62,4 +74,22 @@ public void deleteProduct(@PathVariable("id") UUID id) {
+        }
+
+        GetListResponse responseList = conversionService.convert(listGPR, GetListResponse.class);
+        searchedToPDFsaver.saveSearchedToPdf(responseList);
```
Это уйдет в сервис searchProducts метод

PaleRoom committed:
- fixed Controller
- Savers refactored
- ProductService fixed
- JpaSpecification fixed

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/ProductControllerImpl.java:
```diff
@@ -32,19 +31,15 @@
-        GetListResponse responseList = conversionService.convert(listGPR, GetListResponse.class);
-        return responseList;
+    public List<GetProductResponse> getProducts() {
+        List <GetProductResponse>list =
```
Нужно сразу return без объявления list. А .map и .collect с новой строки

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/ProductControllerImpl.java:
```diff
@@ -76,20 +71,14 @@ public void deleteProduct(@PathVariable("id") UUID id) {
-        searchedToPDFsaver.saveSearchedToPdf(responseList);
-        searchedToXlsSaver.SaveSearchedToXls(responseList);
-        return responseList;
+        List <GetProductResponse>list = productService.searchProducts(immutableSearchProductRequest).stream().map(dto ->
```
77-80 строки не нужны

PaleRoom replied:
Косяк - это я набивал и не убрал.

GUR-ok commented on src/main/java/ru/ncs/DemoShop/repository/specification/ProductSpecification.java:
```diff
@@ -7,29 +7,41 @@
-    else {return (root, query, criteriaBuilder) -> criteriaBuilder.isFalse(root.get(Product_.AVAILABILITY));}
+
+    public static Specification<Product> isAvailable(Boolean availability) {
+        return (root, query, criteriaBuilder) -> {
```
return (root, query, criteriaBuilder) -> {
            if (availability == null) {
                return criteriaBuilder.conjunction();
            } else {
                return criteriaBuilder.equal(root.get(Product_.AVAILABILITY), availability) ;
            }
        };

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/ProductServiceImpl.java:
```diff
@@ -135,19 +137,17 @@ public void increasePrice(double mod) throws InterruptedException {
-        return listDTO;
+        List<ProductDTO> searchedList = productRepository.findAll(specs).stream().map(product ->
+                conversionService.convert(product, ProductDTO.class)).collect(Collectors.toList());
+        searchResultSaver.saveSearchedToPdf(searchedList);
```
вызов двух сейверов давай в трай-кетч обернем и если исключение отловилось, то просто ошибку залогируем. Не хотим запарывать весь запрос, если не смогли сохранить отчет. Вывести инфу по товарам же мы сможем. Из сигнатуры метода убрать throws

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/ProductServiceImpl.java:
```diff
@@ -135,19 +137,17 @@ public void increasePrice(double mod) throws InterruptedException {
+                conversionService.convert(product, ProductDTO.class)).collect(Collectors.toList());
+        searchResultSaver.saveSearchedToPdf(searchedList);
+        searchResultSaver.saveSearchedToXls(searchedList);
+        return searchedList;
```
перед return всегда пустая строка, если метод не состоит из одного лишь return

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/ProductControllerImpl.java:
```diff
@@ -76,20 +71,14 @@ public void deleteProduct(@PathVariable("id") UUID id) {
+                conversionService.convert(dto, GetProductResponse.class)).collect(Collectors.toList());
+//         searchResultSaver.saveSearchedToPdf(list);
+//        searchResultSaver.SaveSearchedToXls(list);
+        return productService.searchProducts(immutableSearchProductRequest).stream().map(dto ->
```
.map .collect с новой строки

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/converter/ProductDtoToGetProductResponseConverter.java:
```diff
@@ -6,7 +6,7 @@
 
 @Component
-final class ConvertToGetProductResponse implements Converter<ProductDTO, GetProductResponse> {
+final class ProductDtoToGetProductResponseConverter implements Converter<ProductDTO, GetProductResponse> {
```
final убрать, сделать public. Остальные конвертеры переименуй плз тоже

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/ProductService.java:
```diff
@@ -21,16 +22,16 @@ public interface ProductService {
     List<ProductDTO> findAll();
 
     @Transactional
```
Я б на методах интерфейса тут transactional не ставил.. т.к. реализация impl может быть разная - в т.ч. и без бд вообще

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/ProductServiceImpl.java:
```diff
@@ -135,19 +137,17 @@ public void increasePrice(double mod) throws InterruptedException {
-            listDTO.add(conversionService.convert(product, ProductDTO.class));
-        }
-        return listDTO;
+        List<ProductDTO> searchedList = productRepository.findAll(specs).stream().map(product ->
```
.map .collect с новой строки

GUR-ok commented on src/main/java/ru/ncs/DemoShop/controller/ProductControllerImpl.java:
```diff
@@ -27,16 +31,19 @@
 public class ProductControllerImpl implements ProductController {
     private final ProductService productService;
     private final ConversionService conversionService;
+    private final SearchResultSaver searchResultSaver;
```
лишнее поле

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/SearchResultSaver.java:
```diff
@@ -0,0 +1,135 @@
+    private void createCell(Row row, int columnCount, Object valueOfCell, CellStyle style) {
+        sheet.autoSizeColumn(columnCount);
+        Cell cell = row.createCell(columnCount);
+        if (valueOfCell instanceof Integer) {
```
А почему нельзя просто cell.setCellValue(valueOfCell.toString())?

PaleRoom replied:
Назначение типов для ячеек Excel

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/SearchResultSaver.java:
```diff
@@ -0,0 +1,135 @@
+            createCell(row, columnCount++, product.getAmountUpdatedAt(), style);
+        }
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yy_MM_dd__HH_mm");
+        FileOutputStream fileOut = new FileOutputStream("Searched_" + LocalDateTime.now().format(formatter) + ".xls");
```
try (FileOutputStream fileOut = new FileOutputStream("Searched_" + LocalDateTime.now().format(formatter) + ".xls")) {
            workbook.write(fileOut);
            workbook.close(); 
        }

GUR-ok commented on src/main/java/ru/ncs/DemoShop/service/SearchResultSaver.java:
```diff
@@ -0,0 +1,135 @@
+@Component
+public class SearchResultSaver {
+    private XSSFSheet sheet;
+    private XSSFWorkbook workbook;
```
Идея предлагает сделать из этого локальную переменную (на 76 строке). И sheet тоже не нужно выносить в поле.

PaleRoom committed:
fixed:
- Converters renamed
- SearchResultSaver
- ProductSpecification (equal added)
- ProductService (saving behavior optimized)

PaleRoom committed:
Merge branch 'develop' into feature/jpaspec

# Conflicts:
#	pom.xml
#	src/main/java/ru/ncs/DemoShop/controller/ProductController.java

PaleRoom committed:
develop merged in JpaSpec, minor fixes
